src/weapon/w_048.c

```diff
@@ -9,7 +9,87 @@ extern s32 g_HandId;
 #define g_Animset w_048_1
 #define g_Animset2 w_048_2
 
-INCLUDE_ASM("weapon/nonmatchings/w_048", EntityWeaponAttack);
+extern WeaponAnimation D_154000_8017AFC0[];
+extern SpriteParts D_154000_8017A040[];
+extern s32 D_154000_8017C3B4;
+
+/* near-duplicate of EntityWeaponAttack_w_049 */
+void EntityWeaponAttack(Entity* self) {
+    s8 animIndex = (self->params & 0x7FFF) >> 8;
+    WeaponAnimation* anim = &D_154000_8017AFC0[animIndex];
+    s32 var_s0;
+
+    self->posX.val = PLAYER.posX.val;
+    self->posY.val = PLAYER.posY.val;
+    self->facingLeft = PLAYER.facingLeft;
+
+    if ((PLAYER.ext.player.anim < anim->frameStart) ||
+        (PLAYER.ext.player.anim >= anim->frameStart + 7) ||
+        g_Player.unk46 == 0) {
+        DestroyEntity(self);
+        return;
+    }
+
+    if (self->step == 0) {
+        SetSpriteBank1(D_154000_8017A040);
+        self->animSet = ANIMSET_OVL(0x10);
+        self->palette = 0x110;
+        self->unk5A = 0x64;
+        if (g_HandId != 0) {
+            self->animSet += 2;
+            self->palette += 0x18;
+            self->unk5A += 2;
+        }
+        self->palette += anim->palette;
+        self->flags = FLAG_UNK_40000 | FLAG_UNK_20000;
+        self->zPriority = PLAYER.zPriority - 2;
+        self->drawMode = DRAW_TPAGE2 | DRAW_TPAGE;
+
+        if (animIndex != 0) {
+            g_api.CreateEntFactoryFromEntity(
+                self, ((g_HandId + 1) << 0xC) + 0x00038, 0);
+            g_api.CreateEntFactoryFromEntity(
+                self, ((g_HandId + 1) << 0xC) + 0x1003A, 0);
+            g_api.CreateEntFactoryFromEntity(
+                self, ((g_HandId + 1) << 0xC) + 0x2003A, 0);
+            g_api.CreateEntFactoryFromEntity(
+                self, ((g_HandId + 1) << 0xC) + 0x3003A, 0);
+            g_api.CreateEntFactoryFromEntity(
+                self, ((g_HandId + 1) << 0xC) + 0x00040, 0);
+        }
+        SetWeaponProperties(self, 0);
+        D_154000_8017C3B4 = 0;
+        self->step++;
+    }
+    self->ext.weapon.anim = PLAYER.ext.player.anim - anim->frameStart;
+    if (PLAYER.animFrameDuration == 1 &&
+        PLAYER.animFrameIdx == anim->soundFrame) {
+        g_api.PlaySfx(anim->soundId);
+    }
+    if (g_api.UpdateUnarmedAnim(anim->frameProps, anim->frames) < 0) {
+        DestroyEntity(self);
+        return;
+    }
+    if (D_154000_8017C3B4 == 0 && PLAYER.animFrameIdx == 1) {
+        inline void create(Entity * self, s32 a, s32 b) {
```

Comment:
This was required to match?

Reply:
Yeah, check the scratch history for another version using vars instead of the inline function. There’s probably something else going on here (with the other create entity calls above) but I haven’t figured out the full pattern.

Reply:
https://decomp.me/scratch/Z9lfD

```c
if ((D_154000_8017C3B4 == 0) && (PLAYER.animFrameIdx == 1)) {
    s32 a, b, c, d, e, f, g, h;
    var_s0 = 0x10;
    if (temp_v0 != 0) {
        var_s0 = 0x20;
    }
    e = ((g_HandId + 1) << 0xC);
    a = ((var_s0 + 0) << 0x10) + 0x38;
    g_api.CreateEntFactoryFromEntity(self, e + a, 0);
    f = ((g_HandId + 1) << 0xC);
    b = ((var_s0 + 1) << 0x10) + 0x3A;
    g_api.CreateEntFactoryFromEntity(self, f + b, 0);
    g = ((g_HandId + 1) << 0xC);
    c = ((var_s0 + 2) << 0x10) + 0x3A;
    g_api.CreateEntFactoryFromEntity(self, g + c, 0);
    h = ((g_HandId + 1) << 0xC);
    d = ((var_s0 + 3) << 0x10) + 0x3A;
    g_api.CreateEntFactoryFromEntity(self, h + d, 0);
}
```

I'm not sure which version is better TBH

Reply:
I think it is a creative idea. If it matches then even better. Only decompiling the PSP version would give us the answer. Until then this looks good